components/autocomplete/index.js

```diff
@@ -247,88 +299,67 @@ export class Autocomplete extends Component {
 	}
 
 	search( event ) {
-		const { open: wasOpen } = this.state;
+		const { open: wasOpen, suppress: wasSuppress } = this.state;
 		const { completers } = this.props;
 		// ensure that the cursor location is unambiguous
 		const container = event.target;
 		const cursor = this.getCursor( container );
 		// look for the trigger prefix and search query just before the cursor location
 		const match = this.findMatch( container, cursor, completers, wasOpen );
 		const { open, query, range } = match || {};
-		// create a regular expression to filter the options
-		const search = open ? new RegExp( escapeRegExp( query ), 'i' ) : /./;
 		// asynchronously load the options for the open completer
 		if ( open && ( ! wasOpen || open.idx !== wasOpen.idx ) ) {
 			this.loadOptions( open.idx );
 		}
+		// create a regular expression to filter the options
+		const search = open ? new RegExp( '(?:\\b|\\s|^)' + escapeRegExp( query ), 'i' ) : /./;
+		// filter the options we already have
+		const filteredOptions = open ? filterOptions( search, this.state[ 'options_' + open.idx ] ) : [];
+		// check if we should still suppress the popover
+		const suppress = open && wasSuppress === open.idx ? wasSuppress : undefined;
 		// update the state
 		if ( wasOpen || open ) {
-			this.setState( { selectedIndex: 0, search, open, query, range } );
-		}
-	}
-
-	getFilteredOptions() {
-		const { maxResults = 10 } = this.props;
-		const { search, open } = this.state;
-		if ( ! open ) {
-			return [];
+			this.setState( { selectedIndex: 0, filteredOptions, suppress, search, open, query, range } );
 		}
-		const options = this.state[ 'options_' + open.idx ] || [];
-
-		const filtered = [];
-		for ( let i = 0; i < options.length; i++ ) {
-			const option = options[ i ];
-
-			// Merge label into keywords
-			let { keywords = [] } = option;
-			if ( 'string' === typeof option.label ) {
-				keywords = [ ...keywords, option.label ];
-			}
-
-			const isMatch = keywords.some( ( keyword ) => search.test( keyword ) );
-			if ( ! isMatch ) {
-				continue;
-			}
-
-			filtered.push( option );
-
-			// Abort early if max reached
-			if ( filtered.length === maxResults ) {
-				break;
-			}
+		// announce the count of filtered options but only if they have loaded
+		if ( open && this.state[ 'options_' + open.idx ] ) {
+			this.announce( filteredOptions );
 		}
-
-		return filtered;
 	}
 
 	setSelectedIndex( event ) {
-		const { open, selectedIndex } = this.state;
-		if ( ! open ) {
+		const { open, suppress, selectedIndex, filteredOptions } = this.state;
+		if ( ! open || filteredOptions.length === 0 ) {
 			return;
 		}
-		const options = this.getFilteredOptions();
-		if ( options.length === 0 ) {
+		if ( suppress === open.idx ) {
+			const { keyCode, ctrlKey, shiftKey, altKey, metaKey } = event;
+			if ( keyCode === SPACE && ctrlKey && ! ( shiftKey || altKey || metaKey ) ) {
```

Comment:
I'm wondering about the `SPACE` check here; if the user moves focus out of the block after dismissing suggestions and then returns later, shouldn't we expect the dismissal to no longer take effect?

![focus-out](https://user-images.githubusercontent.com/1779930/32625153-800e9eca-c559-11e7-80dd-b7c2a25b6400.gif)

Reply:
In a future refactor, to simplify the handling of modifier keys, we might consider adding support to `<KeyboardShortcuts />` to monitor key events of children, mentioned originally in implementation notes #1944 . Made more difficult here by the fact that we manually create the DOM event handlers.

Reply:
Yep, that still hidden after move thing is a bug.

Reply:
The bug should be fixed.